File: C++/Day20.cpp
#include "AH.h"

namespace Day20
{
	enum Mode   { Source, Flip, Conj, Sink };
	enum Signal { Low, High };

	struct GraphNode
	{
		std::string name;
		Mode mode;
		bool onOf; // if this node is a flip flop what state is it in?
		std::map<std::string, bool> inputs;

		GraphNode(const std::string name, const Mode mode) : 
		  name(name), mode(mode), onOf(false), inputs({}) {}
		GraphNode() : name("0"), mode(Mode::Source), onOf(false), inputs({}) {}

		bool allHigh() const;
	};

	bool GraphNode::allHigh() const {
		for (auto [k,v] : inputs) {
			if (!v)
				return false;
		}

		return true;
	}

	struct Graph
	{
		std::vector<std::vector<int>> adj;
		std::vector<GraphNode> gns;
	};
	
	Graph parseInput(const std::vector<std::string> ls)
	{
		const auto n = ls.size() + 1;
		std::vector<std::vector<int>> adj(n, std::vector<int>(n));
		std::vector<GraphNode> gns(n);

		// populate list of names
		int i = 0;
		for (auto l : ls) {
			const auto ps = AH::SplitOnString(l, " -> ");
			std::string name = ps[0];

			auto mode = Mode::Source;
			if (ps[0] != "broadcaster") {
				if (name.at(0) == '&') {
					mode = Mode::Conj;
				} else {
					mode = Mode::Flip;
				}
				name = name.substr(1);
			}
			const GraphNode gn(name, mode);
			gns[i] = gn;
			++i;
		}

		// sneaky hack because i'm lazy fix later, This is the final node
		// gns[n - 1] = GraphNode("output", Mode::Sink);
		gns[n - 1] = GraphNode("rx", Mode::Sink);

		// build adjacency matrix and inputs list
		for (auto i = 0; i < ls.size(); ++i) {
			const auto ps = AH::SplitOnString(ls[i], " -> ");
			std::string name = ps[0];

			if (ps[0] != "broadcaster") {
				name = name.substr(1);
			}
			// vector of nodes this node mapsto
			const auto to = AH::SplitOnString(ps[1], ", ");

			for (auto j = 0; j < gns.size(); ++j) {
				const auto gnName = gns[j].name;
				if (std::find(to.begin(), to.end(), gnName) != to.end()) {
					adj[i][j] = 1;
					gns[j].inputs[name] = false;
				}
			}
		}

		Graph g;
		g.adj = adj;
		g.gns = gns;

		return g;
	}

	typedef std::vector<std::pair<std::string, bool>> Pulses;

	Pulses sendPulse(Graph & g, const Pulses & toProcess, int & low, int & high)
	{	
		// Node <name> has just sent a Pulse <signal> 
		std::vector<std::pair<std::string, bool>> newSignals;
		for (auto [name, signal] : toProcess) {
			// find this node in the graph
			for (auto i = 0; i < g.gns.size(); ++i) {
				auto & node = g.gns[i];
				if (node.name != name) {
					continue;
				}
				// where is signal being sent to?
				const auto adj = g.adj[i];
				for (auto j = 0; j < adj.size(); ++j) {
					if (adj[j] == 0) {
						continue;
					}

					if (signal) {
						high++;
					} else {
						low++;
					}

					// we are transmitting to this node
					auto & child = g.gns[j];
					auto childName = child.name;
					auto childPulse = false; // false = low
					// figure out what sort of Pulse it has to send
					if (child.mode == Mode::Flip) {
						if (!signal) {
							child.onOf = !child.onOf;
							childPulse = child.onOf;
							newSignals.emplace_back(childName, childPulse);
						}
					}
					if (child.mode == Mode::Conj) {
						child.inputs[name] = signal;
						if (child.allHigh()) {
							childPulse = false;
						} else {
							childPulse = true;
						}
						newSignals.emplace_back(childName, childPulse);
					}
					
				}				
			}
		}

		return newSignals;
	}

	int Run(const std::string& filename)
	{
		const auto ls = AH::ReadTextFile(filename);
		auto graph = parseInput(ls);
		
		int lows = 0, highs = 0;
		for (int i = 0; i < 1000; ++i) {
			lows++; // send low signal to broadcaster
			std::vector<std::pair<std:: string, bool>> input = { {"broadcaster", false} };
			while(input.size() > 0) {
				input = sendPulse(graph, input, lows, highs);
			}
		}
	
		AH::PrintSoln(20, lows * highs, 0);

		return 0;
	}

}

File: Go/Day20/Day20.go
//go:build d20
// +build d20

package Day20

import (
	"strings"

	AH "AoC/adventhelper"
)

type Mode int
type Pair struct {
	fst, snd int
}

const (
	Source Mode = iota
	Flip
	Conj
	Sink
)

type GraphNode struct {
	name   string
	mode   Mode
	onOff  bool
	inputs map[string]bool
}

func (gn GraphNode) allHigh() bool {
	for _, v := range gn.inputs {
		if !v {
			return false
		}
	}
	return true
}

func (gn GraphNode) anyHigh() bool {
	for _, v := range gn.inputs {
		if v {
			return true
		}
	}
	return false
}

type Graph struct {
	adj [][]int
	gns []GraphNode
}

func (g Graph) getNode(name string) GraphNode {
	for _, gn := range g.gns {
		if gn.name == name {
			return gn
		}
	}

	return g.gns[0] // should never be hit
}

func parseInput(ss []string) Graph {
	n := len(ss) + 1
	adj := make([][]int, n)
	for i := range adj {
		adj[i] = make([]int, n)
	}
	gns := make([]GraphNode, n)

	for i, s := range ss {
		ps := strings.Split(s, " -> ")
		name := ps[0]

		mode := Source
		if name != "broadcaster" {
			if AH.RuneAt(name, 0) == '&' {
				mode = Conj
			} else {
				mode = Flip
			}
			name = AH.TrimFirstRune(name)
		}
		mm := make(map[string]bool)
		gn := GraphNode{name: name, mode: mode, onOff: false, inputs: mm}
		gns[i] = gn
	}

	mn := make(map[string]bool)
	gns[n-1] = GraphNode{"rx", Sink, false, mn}

	for i := range ss {
		ps := strings.Split(ss[i], " -> ")
		name := ps[0]
		if ps[0] != "broadcaster" {
			name = AH.TrimFirstRune(name)
		}
		to := strings.Split(ps[1], ", ")
		for j := range gns {
			gnName := gns[j].name
			if AH.ContainsStr(to, gnName) {
				adj[i][j] = 1
				gns[j].inputs[name] = false
			}
		}
	}

	g := Graph{adj, gns}

	return g
}

type Pulse struct {
	name  string
	onOff bool
}

func sendPulse(g Graph, toProcess []Pulse, low *int, high *int) []Pulse {
	newSignals := make([]Pulse, 0)
	for _, s := range toProcess {
		name := s.name
		signal := s.onOff

		for i := range g.gns {
			node := g.gns[i]
			if node.name != name {
				continue
			}

			adj := g.adj[i]
			for j := range adj {
				if adj[j] == 0 {
					continue
				}

				if signal {
					*high++
				} else {
					*low++
				}

				childName := g.gns[j].name
				childPulse := false

				if g.gns[j].mode == Flip {
					if !signal {
						g.gns[j].onOff = !g.gns[j].onOff
						childPulse = g.gns[j].onOff
						newSignals = append(newSignals, Pulse{childName, childPulse})
					}
				} else if g.gns[j].mode == Conj {
					g.gns[j].inputs[name] = signal
					if g.gns[j].allHigh() {
						childPulse = false
					} else {
						childPulse = true
					}
					newSignals = append(newSignals, Pulse{childName, childPulse})
				}
			}
		}
	}

	return newSignals
}

func Run() {
	ls, _ := AH.ReadStrFile("../input/input20.txt")
	graph := parseInput(ls)

	lows, highs := 0, 0
	for i := 0; i < 1000; i++ {
		lows++
		input := []Pulse{{"broadcaster", false}}
		for len(input) > 0 {
			input = sendPulse(graph, input, &lows, &highs)
		}
	}

	part1 := lows * highs

	n := graph.getNode("cl")
	cycles := make(map[string]Pair)
	for k, _ := range n.inputs {
		cycles[k] = Pair{0, 0}
	}

	for i := 0; i < 10000; i++ {
		input := []Pulse{Pulse{"broadcaster", false}}

		for len(input) > 0 {
			input = sendPulse(graph, input, &lows, &highs)

			n := graph.getNode("cl")
			if n.anyHigh() {
				for name, v := range n.inputs {
					if v {
						p := cycles[name]
						if i == p.snd {
							break
						}
						cycles[name] = Pair{p.snd, i}
						break
					}
				}
			}
		}
	}

	part2 := 1
	for _, v := range cycles {
		part2 = AH.LCM(part2, v.snd-v.fst)
	}

	AH.PrintSoln(20, part1, part2)

	return
}
